Frontend-unit-tests-guide.md

```diff
@@ -1,409 +1,484 @@
-  });
-```
+* For all hardcoded values, explain where the values come from in comments.
 
-### Structure of a Test
+* Write the expectations in order, so that each test has a coherent story. That makes the test easier to follow.
 
+#### Similar tests should have similar checks
 
-*   Keep the body of the test clean and clear. Having too many big blocks of data obscures what we are trying to check. In a case where big blocks of data is necessary to test the code, add a verbal or visual explanation (as [here](https://github.com/oppia/oppia/blob/3fa63711440b4320350c48dd63ffd8b163d07143/core/templates/components/graph-services/graph-layout.service.spec.ts#L772-L784)) through comments.
-*   An ideal test should include both pre-checks and post-checks.
-*   Do not include unnecessary data in the test.
-    *   **Example** : if you are testing that the X property of a dict is modified by a function, then only check the X property. Do not include other properties in the expected variable.
-    *   Instead of expectedDict, make an expectedDictXValues list, and check if the values are the same as expected in the returned dict.
-*   Leave comments to explain the steps in the test whenever things get complicated. This helps the reader understand what your test does. Ideally, a reader should be able to understand the tests without reading the code file.
-*   Add visual explanations if possible ([example](https://github.com/oppia/oppia/blob/1013c9a33af0179ad394088c08ef18ed2f03bd64/core/templates/components/graph-services/graph-layout.service.spec.ts#L28-L57)).
-*   For all hardcoded values -- explain how you got that (how it was calculated).
-*   For similar tests, you should have parity in the checks
-    *   **Example**: If in a test you check that a spy was called when a certain condition was satisfied, then also check that the spy was not called when that condition was not satisfied.
-*   Write the expectations in order, so that each test has a coherent story. That makes the test easier to follow.
-*   Always validate an external side effect.
-    *   **Example**: In [this](https://github.com/oppia/oppia/blob/c714eeb320e0ebf9fe400403569428933658d3f1/core/templates/components/button-directives/create-activity-button.component.ts#L49-L56) function, the call to siteAnalyticsService is an external side effect.
+For example, if in a test you check that a spy was called when a certain condition was satisfied, then also check that the spy was not called when that condition was not satisfied.
 
-### Contacts
+#### Validate external side-effects
 
-If you have any questions about the above, you can contact any of the people below:
-* @aishwary023 - aishwary.saxena.min19@iitbhu.ac.in
-* @gp201 - praneethg2001@gmail.com
-* @Radesh-kumar - imradesh@gmail.com
+An external side-effect is an effect that we expect from running the function but that isn't part of the outputs. For example, consider this pseudocode:
+
+```text
+function logIn(username, password) {
```

Comment:
I think we use `login` in general, right? Ditto below in setLoginCookie.

Reply:
done